discovery/zookeeper/metrics.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2025 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zookeeper
+
+import (
+	"errors"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/prometheus/prometheus/discovery"
+)
+
+type zookeeperMetrics struct {
+	// The total number of ZooKeeper failures.
+	failureCounter prometheus.Counter
+	// The current number of Zookeeper watcher goroutines.
+	numWatchers prometheus.Gauge
+}
+
+// Create and register metrics.
+func newDiscovererMetrics(reg prometheus.Registerer, _ discovery.RefreshMetricsInstantiator) discovery.DiscovererMetrics {
+	m := &zookeeperMetrics{
+		failureCounter: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: "prometheus",
+			Subsystem: "treecache",
+			Name:      "zookeeper_failures_total",
+			Help:      "The total number of ZooKeeper failures.",
+		}),
+		numWatchers: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "prometheus",
+			Subsystem: "treecache",
+			Name:      "watcher_goroutines",
+			Help:      "The current number of watcher goroutines.",
+		}),
+	}
+	// For historical reasons, both ServerSet and Nerve SD share the same zookeeper metrics.
+	// To not cause double registration problems, if both SD mechanisms are instantiated with
+	// the same registry, we are handling the AlreadyRegisteredError accordingly below.
```

Comment:
I suppose that the metrics should normally get unregistered if the config file is reloaded, but in this case it won't happen? It would be nice if the two SDs indeed use different metrics one day to make sure such edge cases are handled ok. But for now it should be ok to use `AlreadyRegisteredError`.

Reply:
Yes, we should definitely aim for separating those metrics one day.

Unregistering is a bit of a problem, I guess, once both ServerSet and Nerve SD are used at the same time. Maybe let's not do the unregistering in this case (but leave a comment why that is the case and that we should one day separate the metrics for both SDs).